README.md
# Java Fitch
Basic Java Unit Testing framework used by Black Fitch


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## 0.1.0 - 2017-05-28
- Split project from Black Fitch

### Added
- Colours for headers and test results
- Logging class
- Line numbers with failed asserts

### Changed
- Split the Runner into JavaFitch and JavaFitchRunner
- Change license from MIT to GPL-3.0

### Fixed
- Not collecting the result from test suites


JLogger.java
public class JLogger {

    public static final int DEBUG = 0;
    public static final int ERROR = 1;
    public static final int FATAL = 2;
    public static final int INFO = 3;
    public static final int WARN = 4;

    public static int logLevel = 0;

    public static void log(int level, int[][] msg) {

        String res = "";

        for (int i = 0; i < msg.length; i++) {
            for (int j = 0; j < msg.length; j++) {
                res += Integer.toString(msg[i][j]) + " ";
            }
            res += "\n";
        }

        log(level, "\n" + res);
    }

    public static void log(int level, boolean msg) {
        log(level, Boolean.toString(msg));
    }

    public static void log(int level, int msg) {
        log(level, Integer.toString(msg));
    }

    public static void log(int level, String msg) {

        if (logLevel <= level) {
            int callersLineNumber = Thread.currentThread().getStackTrace()[2].getLineNumber();
            logTemplate(level, msg, callersLineNumber);
        }
    }

    public static void logTemplate(int level, String msg, int line) {
        // TODO: Use a map?

        String typeName = "";

        if (level == DEBUG) {
            typeName = "DEBUG";
        } else if (level == ERROR) {
            // NOTE: Not all terminals support
            typeName = "\u001B[31mERROR\u001B[39;49m";
        } else if (level == WARN) {
            typeName = "\u001B[31mWARN\u001B[39;49m";
        }

        System.out.println("[" + typeName + "][" + line + "] " + msg);
    }
}


JavaFitch.java
/*
This class manages test suites and collects their results
*/

import java.util.*;

public class JavaFitch {

	public static final String VERSION = "v0.1.0";

	public static ArrayList<JavaFitchUnitTest> testList = new ArrayList<JavaFitchUnitTest>();

	public static String TITLE_COLOUR = "\u001B[38;5;44m";
	public static String TITLE_COLOUR_2 = "\u001B[38;5;202m";
	public static String GREEN = "\u001B[32m";
	public static String RED = "\u001B[31m";
	public static String RESET = "\u001B[39;49m";

	static int testsCount = 0;
	static int testsPassed = 0;
	static int assertsCount = 0;
	static int assertsPassed = 0;
	static boolean result = true;

	public static void testWrapper(JavaFitchUnitTest test) {
		printHeader(test.name);
		result = result & test.run();

		testsCount += test.testsCount;
		testsPassed += test.testsPassed;
		assertsCount += test.assertsCount;
		assertsPassed += test.assertsPassed;
	}

	public static void runAllTests() {
		for (int i = 0; i < testList.size(); i++) {
			testWrapper(testList.get(i));
		}

		for (int i = 0; i < testList.size(); i++) {
			printFooter(testList.get(i).name, testList.get(i).testResult,
			testList.get(i).testsPassed, testList.get(i).testsCount,
			testList.get(i).assertsPassed, testList.get(i).assertsCount);
		}

		printFooter("OVERALL", result, testsPassed, testsCount, assertsPassed, assertsCount);
	}

	public static void addSuite(JavaFitchUnitTest test) {
		testList.add(test);
	}

	public static void printHeader(String title) {
		System.out.println("XXXXXXXXX SUITE: " + TITLE_COLOUR + title + RESET + " XXXXXXXXXX");
	}

	public static void printFooter(String title, boolean result, int testsCount, int testsPassed, int assertsCount, int assertsPassed) {
		System.out.println("________ SUMMARY for " + TITLE_COLOUR + title + RESET + " _________");
		System.out.println(resultToString(result));
		System.out.println("tests: " + testsPassed + " passed | " + testsCount);
		System.out.println("assertions: " + assertsPassed + " passed | " + assertsCount);
		System.out.println("_____________________________________________");


	}

	public static String resultToString(boolean value) {

		String result = "";

		if (value == true) {
			result = GREEN + "PASS" + RESET;
		} else {
			result = RED + "FAIL" + RESET;
		}

		return result;
	}
}


tests_Task1.java
/*
USER FILE

Example test suite
*/

public class tests_Task1 extends JavaFitchUnitTest{

	public tests_Task1(String n) {
		super(n);
	}

	public boolean run() {

		boolean result = true;
		result = result & test_template("Case: Single Node");

		return result;
	}


	public boolean test_template(String title) {
		// HEADER
		printHeader(title);
		boolean result = true;
		boolean subResult = true;
		// HEADER END

		// =========================================================
		// TEST CASE START										VVVV
		subResult = start_test_case("Creating a new BPlusTree with M = 5");
		// ==========================================================

		subResult = subResult & assertEquals("123", "-1");

		// =========================================================
		result = end_test_case(result, subResult);
		// TEST CASE END										^^^^
		// =========================================================


		// FOOTER
		printFooter(title, result);
		return result;
		// FOOTER END
	}


}


JavaFitchLogger.java
public class JavaFitchLogger {

    public static final int DEBUG = 0;
    public static final int ERROR = 1;
    public static final int FATAL = 2;
    public static final int INFO = 3;
    public static final int WARN = 4;

    public int logLevel = 0;
    public String name = "";

    public JavaFitchLogger(int defaultLevel) {
        logLevel = defaultLevel;
    }

    public JavaFitchLogger(String n, int defaultLevel) {
        logLevel = defaultLevel;
        name = n;
    }

    public void log(int level, String msg) {
        if (logLevel <= level) {
            int callersLineNumber = Thread.currentThread().getStackTrace()[2].getLineNumber();
            logTemplate(level, msg, callersLineNumber);
        }
    }

    public void logTemplate(int level, String msg, int line) {
        // TODO: Use a map?

        String typeName = "";

        if (level == DEBUG) {
            typeName = "DEBUG";
        } else if (level == ERROR) {
            typeName = "\u001B[31mERROR\u001B[39;49m";
        }

        System.out.println("[" + typeName + "][" + line + "][" + name + "] " + msg);
    }
}


JavaFitchUnitTest.java
/*
Defines the base functions for a test suite
*/

public class JavaFitchUnitTest {

	public static String TITLE_COLOUR = "\u001B[38;5;44m";
	public static String TITLE_COLOUR_2 = "\u001B[38;5;202m";
	public static String GREEN = "\u001B[32m";
	public static String RED = "\u001B[31m";
	public static String RESET = "\u001B[39;49m";

	public int testsPassed = 0;
	public int testsCount = 0;
	public int assertsPassed = 0;
	public int assertsCount = 0;
	public int assertsPassedSubTest = 0;
	public int assertsCountSubTest = 0;
	public boolean testResult = true;
	public String name = "";

	public JavaFitchUnitTest(String n) {
		name = n;
	}

	public boolean run() {
		System.out.println("JavaFitchUnitTest::run() not implemented!");

		return false;
	}

	public void printHeader(String title) {
		title = TITLE_COLOUR_2 + title + RESET;
		assertsCountSubTest = 0;
		assertsPassedSubTest = 0;

		System.out.println("\t=========================");
		System.out.println("\t" + title);
		System.out.println("\t=========================");

	}

	public boolean start_test_case(String title) {
		printSubTest(title);
		return true;
	}

	public boolean end_test_case(boolean result, boolean subResult) {
		printSubTestFooter(subResult);
		testResult = testResult & subResult;
		return result & subResult;
	}

	public void printFooter(String title, boolean result) {
		testsCount++;

		if (result == true) {
			testsPassed++;
		}

		System.out.println("\t_____ SUMMARY: " + title + " _____");
		//System.out.print("\t" + title + " Summary: ");
		System.out.println("\t" + resultToString(result));
		System.out.println("\tassertions: " + assertsPassedSubTest + " passed | " + assertsCountSubTest);
		System.out.println("\t_________________________");
	}

	public void printSubTest(String title) {

		System.out.println("\t\t-----------------------");
		System.out.print("\t\t" + title + ": ");
	}

	public void printSubTestFooter(boolean value) {
		if (value != false) {
			System.out.println(resultToString(value));
			System.out.println("\t\t-----------------------");
		} else {
			System.out.println("\n\t\t-----------------------");
		}

	}

	public String resultToString(boolean value) {

		String result = "";

		if (value == true) {
			result = GREEN + "PASS" + RESET;
		} else {
			result = RED + "FAIL" + RESET;
		}

		return result;
	}

	public <T> boolean assertEquals(T expected, T received) {
		int tracer = Thread.currentThread().getStackTrace()[2].getLineNumber();

		assertsCount++;
		assertsCountSubTest++;

		if (received.equals(expected)) {
			assertsPassed++;
			assertsPassedSubTest++;
			return true;
		} else {
			System.out.println();
			System.out.println(RED + "++++ ASSERT FAIL on line " + tracer + " ++++" + RESET);
			System.out.println("Expected: \n" + expected.toString());
			System.out.println("Received: \n" + received.toString() + "\n");
			System.out.println(RED + "++++++++++++++++++++++" + RESET);

			return false;
		}
	}

	public <T> boolean assertNull(T received, int tracer) {
		tracer++;
		assertsCount++;
		assertsCountSubTest++;

		if (received == null) {
			assertsPassed++;
			assertsPassedSubTest++;
			return true;
		} else {
			System.out.println();
			System.out.println(RED + "++++ ASSERT FAIL on line " + tracer + " ++++" + RESET);
			System.out.println("Expected: \n" + "null");
			System.out.println("Received: \n" + received.toString() + "\n");
			System.out.println(RED + "++++++++++++++++++++++" + RESET);

			return false;
		}
	}

	public <T> boolean assertNull(T received) {

		assertsCount++;
		assertsCountSubTest++;

		if (received == null) {
			assertsPassed++;
			assertsPassedSubTest++;
			return true;
		} else {
			System.out.println();
			System.out.println(RED + "++++ ASSERT FAIL ++++" + RESET);
			System.out.println("Expected: \n" + "null");
			System.out.println("Received: \n" + received.toString() + "\n");
			System.out.println(RED + "++++++++++++++++++++++" + RESET);

			return false;
		}
	}
}
